### DISCUSSION TITLE
👋 Welcome to GoogleReader API Integration Testing

### DISCUSSION THREAD

**Discussion by Ashinch** [ReadYouApp/ReadYou#542] [Announcements]
In the latest main branch, the integration of the GoogleReader API has been merged, and we are looking for thorough validation before the next release.

🤔 How to Participate?

⚠️ Please make sure to back up your data before proceeding.

Choose the latest workflow from the [GitHub Actions page](https://github.com/Ashinch/ReadYou/actions?query=branch%3Amain+event%3Apush) and download the artifacts specific to the GitHub version from the workflow's summary page (You may require logging in to GitHub).

🐛 Below are known issues that will be addressed in subsequent commits:

1. Incompatibility with Inoreader.
2. Unable to specify a new group when adding a subscription.
3. ~Marking a feed as read does not take effect.~
4. ~Incompatibility with Miniflux.~
5. Import from OPML file.

**Comment by Alkarex** [comment]
Congrats! Very promising already 👍🏻 

A few notes after a brief test:

1. It would be nice to be able to control how many articles to sync (including some high values such as 10k)
2. It looks like articles are re-downloaded during sync, which is very slow and wasting resources (battery, network). Please see [synchronisation strategy tips](https://freshrss.github.io/FreshRSS/en/developers/06_GoogleReader_API.html#synchronisation-strategy) (cf. https://github.com/Ashinch/ReadYou/issues/221):
    * [Synchronisation recommendation by Alkarex](https://github.com/FreshRSS/FreshRSS/issues/2566#issuecomment-541317776)
    * [The Right Way to Sync by BazQux](https://github.com/bazqux/bazqux-api#user-content-the-right-way-to-sync)
3. When scrolling down a long list of articles, it would be nice to be able to mark as read only the ones already passed and not the rest of the list, for instance by doing a long click on an article to "mark above articles as read".
4. The app does not seem keep track of articles marked as read or starred while offline:
    a. Open the app and sync while connected
    b. Set phone to airplane mode
    c. Read some articles, and star some articles
    d. Close app
    e. Disable airplane mode on phone to get connectivity back
    f. Open the app
    g. The read and star status of articles changed while offline is lost (articles are still unread, with no star)
5. When "pulling to refresh" from within a feed, it refreshes everything instead of getting only articles from that feed. Assuming there are many more articles on the server than in the local application, that prevents getting more (unread) articles for the current feed.

Feel free to ping me for anything related to the Google Reader API and/or the [FreshRSS back-end](https://github.com/FreshRSS/FreshRSS). I can also offer a test account populated with many feeds and articles to test performance.

**Reply by Ashinch** [reply]
Hi, @Alkarex 

Thanks to FreshRSS for its support of the community ecosystem!

> 1. It would be nice to be able to control how many articles to sync (including some high values such as 10k)

This is a great suggestion.

> 3. When scrolling down a long list of articles, it would be nice to be able to mark as read only the ones already passed and not the rest of the list, for instance by doing a long click on an article to "mark above articles as read".

Due to the nature of Compose, implementing the scrolling mark feature is challenging and may require further research. However, the swipe mark feature is available now.

> 2. It looks like articles are re-downloaded during sync, which is very slow and wasting resources (battery, network). 
> 5. When "pulling to refresh" from within a feed, it refreshes everything instead of getting only articles from that feed. Assuming there are many more articles on the server than in the local application, that prevents getting more (unread) articles for the current feed.

Okay, I referenced Reeder's sync mechanism with the GoogleReader API. As far as I understand, Reeder seems to auto alternate between full sync and incremental sync, I think I can also make some optimizations (If you have more details about the Reeder sync mechanism, please let me know too).

> 4. The app does not seem keep track of articles marked as read or starred while offline:

A novel perspective! This would require a refactoring of the network request buffer queue before we can achieve it.

**Reply by Alkarex** [reply]
Regarding point 2 (*articles are re-downloaded during sync*),  this is more or less what can already be seen in https://github.com/FreshRSS/FreshRSS/issues/2566#issuecomment-541317776 and I can provide additional traces of the calls to perform in other situations if needed.

Regarding *"pulling to refresh" from within a feed*, it is something else, and is easier, as it is just about adding a filter `s=` for the target stream / feed. (I can provide the exact call / trace if wanted)

To check what other apps are doing, it is possible to just log the API requests received by the FreshRSS back-end.
For instance by uncommenting the following lines and checking `./FreshSS/data/users/_/log_api.txt` (I can also help with that):
https://github.com/FreshRSS/FreshRSS/blob/4a0e8a70580c96ec2aeed2cf44bd79bc302676f8/p/api/greader.php#L1003-L1004

The best reference I have seen [is News+](https://github.com/FreshRSS/FreshRSS/blob/edge/README.md#apis--native-apps), which has not been updated in a decade (still available as manual install), but which went through several iterations with improved synchronisation strategies, for ending with what I believe is the best possible strategy, so this is what I recommend to take as a reference.

**Reply by Alkarex** [reply]
P.S. Read You added to FreshRSS readme: https://github.com/FreshRSS/FreshRSS/blob/edge/README.md#apis--native-apps

**Reply by mbestavros** [reply]
@Ashinch I'll echo @Alkarex's point 2. Fetching the entire article database is not sustainable, especially as more feeds get added.

IMO, that's a must-fix before release to the public. The current sync experience feels very janky.

**Reply by Ashinch** [reply]
> I can also offer a test account populated with many feeds and articles to test performance.

Hi, @Alkarex 

Could you please offer me a test account populated with many feeds and articles to test performance? 

My email: `Ashinch@outlook.it`.

**Reply by Ashinch** [reply]
> @Ashinch I'll echo @Alkarex's point 2. Fetching the entire article database is not sustainable, especially as more feeds get added.
> 
> IMO, that's a must-fix before release to the public. The current sync experience feels very janky.

@mbestavros 

True, I'm making some changes.

**Reply by Alkarex** [reply]
> Could you please offer me a test account populated with many feeds and articles to test performance?

Done! I have just sent you an e-mail

**Reply by Ashinch** [reply]
> Done! I have just sent you an e-mail

@Alkarex Thank you for your help. I've received your email and I'm ready to start optimizing the synchronization strategy.

**Reply by Ashinch** [reply]
Hi, @Alkarex 

I made some changes in #570, to which I've attached some instructions.

**Comment by Alkarex** [comment]
Something more: an important feature of the Google Reader API is the ability for users to add user labels (also called tags) to the articles, in order to keep them / re-share them.

It would be nice to support that in Read You. On the back-end side, FreshRSS supports the possibility to manually tag articles as well as to auto-tag articles based on different criteria, which is a popular feature.

An article may appear in a feed as well as in multiple tags (and potentially also in the star / favourite list).

Here is a screenshot from News+ as an illustration:
![Screenshot_20200229-134519_News+](https://user-images.githubusercontent.com/1008324/75607792-e9db2b00-5afa-11ea-9a29-205bb5ce0b0f.png)

Do not hesitate to ask if if you need any additional documentation / examples.

**Reply by Ashinch** [reply]
Hi, @Alkarex 

Thank you for your reply.

About tags: #546

I have another question. How to ensure the consistency of reading on multiple devices according to the synchronization policy in #221? Maybe users don't just read on the ReadYou client. If ReadYou filters with `ot` (last sync time) when pulling the items, then we may miss some changes made earlier-items on other clients.

**Reply by Alkarex** [reply]
Indeed, it is expected that users read both from the Web interface and from one or more apps.

With the `ot` as [suggested in point 3](https://github.com/FreshRSS/FreshRSS/issues/2566#issuecomment-541317776), you will get the new unread articles and their content, excluding the articles your app already has. This request will skip new articles already read elsewhere (hence giving priority to unread articles). (Unread articles can be fetched on demand when desired - see also my suggestion 5 higher up).

You do get status modifications (changes) of more articles in a cheaper way thanks to query 4 (and for stared / favourites with queries 5, 6, 7).

This works very well and is very efficient (it takes only a fraction of the time taken by some other apps).
I hope this clarifies, and I can happily provide more info if needed.
